components/fxa-client/src/oauth.rs

```diff
@@ -167,25 +166,49 @@ impl FirefoxAccount {
     /// Fetch an OAuth code for a particular client using a session token from the account state.
     /// This method doesn't support OAuth public clients at this time.
     ///
-    /// * `client_id` - OAuth client id.
-    /// * `scopes` - Space-separated list of requested scopes.
-    /// * `state` - OAuth state.
-    /// * `access_type` - Type of OAuth access, can be "offline" and "online.
+    /// * `auth_params` Authentication parameters
     pub fn authorize_code_using_session_token(
         &self,
-        client_id: &str,
-        scope: &str,
-        state: &str,
-        access_type: &str,
+        auth_params: AuthorizationParameters,
     ) -> Result<String> {
         let session_token = self.get_session_token()?;
+        let (code_challenge, code_challenge_method, keys_jwe) = if let Some(key_params) =
+            auth_params.key_params
+        {
+            let keys_jwk = base64::decode_config(key_params.keys_jwk, base64::URL_SAFE_NO_PAD)?;
+            let keys_jwk = String::from_utf8(keys_jwk)?;
+            let requested_scopes: HashSet<&str> = auth_params.scope.split_whitespace().collect();
+            let data: HashMap<String, ScopedKey> = self
+                .state
+                .scoped_keys
+                .iter()
+                .filter(|(scope, _)| requested_scopes.contains(&(*scope).as_ref()))
+                .map(|(key, value)| (key.clone(), value.clone()))
+                .collect();
+            let data = serde_json::to_string(&data)?;
+            let client_keys_flow = ScopedKeysFlow::with_random_key()?;
+            (
+                Some(key_params.code_challenge),
+                Some(key_params.code_challenge_method),
+                Some(client_keys_flow.encrypt_keys_jwe(&keys_jwk, data.as_bytes())?),
+            )
+        } else {
+            (None, None, None)
```

Comment:
It's going to be a lot of churn in your PR, but I think one benefit of separating the PKCE parms from the JWK will be that it simplifies the control-flow code here.